engine/access/rest/websockets/controller.go

```diff
@@ -3,219 +3,416 @@ package websockets
+			continue
+		}
 	}
-	return message, nil
 }
 
-func (c *Controller) parseAndValidateMessage(message json.RawMessage) (models.BaseMessageRequest, interface{}, error) {
+func (c *Controller) parseAndValidateMessage(ctx context.Context, message json.RawMessage) error {
 	var baseMsg models.BaseMessageRequest
 	if err := json.Unmarshal(message, &baseMsg); err != nil {
-		return models.BaseMessageRequest{}, nil, fmt.Errorf("error unmarshalling base message: %w", err)
+		return fmt.Errorf("error unmarshalling base message: %w", err)
 	}
 
 	var validatedMsg interface{}
 	switch baseMsg.Action {
-	case "subscribe":
+	case models.SubscribeAction:
 		var subscribeMsg models.SubscribeMessageRequest
 		if err := json.Unmarshal(message, &subscribeMsg); err != nil {
-			return baseMsg, nil, fmt.Errorf("error unmarshalling subscribe message: %w", err)
+			return fmt.Errorf("error unmarshalling subscribe message: %w", err)
 		}
-		//TODO: add validation logic for `topic` field
 		validatedMsg = subscribeMsg
 
-	case "unsubscribe":
+	case models.UnsubscribeAction:
 		var unsubscribeMsg models.UnsubscribeMessageRequest
 		if err := json.Unmarshal(message, &unsubscribeMsg); err != nil {
-			return baseMsg, nil, fmt.Errorf("error unmarshalling unsubscribe message: %w", err)
+			return fmt.Errorf("error unmarshalling unsubscribe message: %w", err)
 		}
 		validatedMsg = unsubscribeMsg
 
-	case "list_subscriptions":
+	case models.ListSubscriptionsAction:
 		var listMsg models.ListSubscriptionsMessageRequest
 		if err := json.Unmarshal(message, &listMsg); err != nil {
-			return baseMsg, nil, fmt.Errorf("error unmarshalling list subscriptions message: %w", err)
+			return fmt.Errorf("error unmarshalling list subscriptions message: %w", err)
 		}
 		validatedMsg = listMsg
 
 	default:
 		c.logger.Debug().Str("action", baseMsg.Action).Msg("unknown action type")
-		return baseMsg, nil, fmt.Errorf("unknown action type: %s", baseMsg.Action)
+		return fmt.Errorf("unknown action type: %s", baseMsg.Action)
 	}
 
-	return baseMsg, validatedMsg, nil
+	c.handleAction(ctx, validatedMsg)
+	return nil
 }
 
-func (c *Controller) handleAction(ctx context.Context, message interface{}) error {
+func (c *Controller) handleAction(ctx context.Context, message interface{}) {
 	switch msg := message.(type) {
 	case models.SubscribeMessageRequest:
 		c.handleSubscribe(ctx, msg)
 	case models.UnsubscribeMessageRequest:
 		c.handleUnsubscribe(ctx, msg)
 	case models.ListSubscriptionsMessageRequest:
 		c.handleListSubscriptions(ctx, msg)
-	default:
-		return fmt.Errorf("unknown message type: %T", msg)
 	}
-	return nil
 }
 
 func (c *Controller) handleSubscribe(ctx context.Context, msg models.SubscribeMessageRequest) {
-	dp, err := c.dataProviderFactory.NewDataProvider(ctx, msg.Topic, msg.Arguments, c.communicationChannel)
+	// register new provider
+	provider, err := c.dataProviderFactory.NewDataProvider(ctx, msg.Topic, msg.Arguments, c.multiplexedStream)
```

Comment:
Initially, we thought that the use of a callback here would prevent a possible write to a closed channel but it turned out it would not. So, it doesn't matter what approach to use